chat.go

```diff
@@ -61,8 +62,58 @@ type ChatCompletionRequest struct {
 type FunctionDefine struct {
 	Name        string `json:"name"`
 	Description string `json:"description,omitempty"`
-	// it's required in function call
-	Parameters *FunctionParams `json:"parameters"`
+	// Parameters is a JSONSchema object describing the function. Generate it via another library.
+	Parameters    *FunctionParams `json:"-"`
+	ParametersRaw json.RawMessage `json:"-"`
+}
+
+func (fd FunctionDefine) MarshalJSON() ([]byte, error) {
```

Comment:
Could somebody please review this marshal/unmarshal code? I barely understand it, if we are going to keep Parameters and ParametersRaw then this code _needs_ to be correct.

Reply:
It looks okay to me. The custom marshalling is almost exactly the same as [in this article](https://medium.com/picus-security-engineering/custom-json-marshaller-in-go-and-common-pitfalls-c43fa774db05).

The only thing I would change would be to rename `FunctionDefine` to `FunctionDefinition`...